File: javascript/src/memory.ts
/**
 * BlackRoad Memory API
 */

import type { BlackRoadClient } from './client';
import type { MemoryEntry, MemoryQueryOptions, LogMemoryOptions, Stats } from './types';

export class MemoryAPI {
  constructor(private client: BlackRoadClient) {}

  /**
   * Log an entry to the memory system.
   */
  async log(options: LogMemoryOptions): Promise<MemoryEntry> {
    return this.client.post<MemoryEntry>('/memory', {
      action: options.action,
      entity: options.entity,
      details: options.details,
      tags: options.tags,
      metadata: options.metadata,
    });
  }

  /**
   * Query memory entries.
   */
  async query(options: MemoryQueryOptions = {}): Promise<MemoryEntry[]> {
    const params: Record<string, unknown> = {
      limit: options.limit || 100,
      offset: options.offset || 0,
    };

    if (options.search) params.q = options.search;
    if (options.action) params.action = options.action;
    if (options.entity) params.entity = options.entity;
    if (options.tags) params.tags = options.tags.join(',');
    if (options.since) params.since = options.since.toISOString();
    if (options.until) params.until = options.until.toISOString();

    const response = await this.client.get<{ entries: MemoryEntry[] }>('/memory', params);
    return response.entries || [];
  }

  /**
   * Get a specific memory entry by hash.
   */
  async get(entryHash: string): Promise<MemoryEntry> {
    return this.client.get<MemoryEntry>(`/memory/${entryHash}`);
  }

  /**
   * Get recent memory entries.
   */
  async recent(limit: number = 50): Promise<MemoryEntry[]> {
    return this.query({ limit });
  }

  /**
   * Get agent state.
   */
  async agentState(agentId: string): Promise<Record<string, unknown>> {
    return this.client.get<Record<string, unknown>>(`/memory/agents/${agentId}/state`);
  }

  /**
   * Sync agent state.
   */
  async syncState(agentId: string, state: Record<string, unknown>): Promise<{ synced: boolean }> {
    return this.client.post<{ synced: boolean }>(`/memory/agents/${agentId}/state`, state);
  }

  /**
   * Broadcast a message.
   */
  async broadcast(type: string, payload: string): Promise<{ broadcast_id: string }> {
    return this.client.post<{ broadcast_id: string }>('/memory/broadcast', { type, payload });
  }

  /**
   * Share a "Today I Learned" entry.
   */
  async til(category: string, learning: string): Promise<MemoryEntry> {
    return this.log({
      action: 'til',
      entity: category,
      details: learning,
      tags: ['til', category],
    });
  }

  /**
   * Get memory statistics.
   */
  async stats(): Promise<Stats> {
    return this.client.get<Stats>('/memory/stats');
  }

  /**
   * Verify hash chain integrity.
   */
  async verifyChain(startHash?: string): Promise<{ valid: boolean; checked: number }> {
    const params = startHash ? { start: startHash } : {};
    return this.client.get<{ valid: boolean; checked: number }>('/memory/verify', params);
  }
}

File: rust/src/memory.rs
use crate::client::BlackRoadClient;
use crate::errors::Error;
use crate::types::{LogMemoryOptions, MemoryEntry, MemoryQueryOptions, Stats, VerifyChainResult};
use serde::Deserialize;
use std::collections::HashMap;

/// API for memory operations.
#[derive(Debug, Clone)]
pub struct MemoryAPI {
    client: BlackRoadClient,
}

#[derive(Deserialize)]
struct EntriesResponse {
    entries: Vec<MemoryEntry>,
}

#[derive(Deserialize)]
struct BroadcastResponse {
    broadcast_id: String,
}

impl MemoryAPI {
    pub(crate) fn new(client: BlackRoadClient) -> Self {
        Self { client }
    }

    /// Logs a new memory entry.
    pub async fn log(&self, opts: LogMemoryOptions) -> Result<MemoryEntry, Error> {
        self.client.post("/memory", &opts).await
    }

    /// Queries memory entries.
    pub async fn query(&self, opts: Option<MemoryQueryOptions>) -> Result<Vec<MemoryEntry>, Error> {
        let mut params = HashMap::new();
        params.insert("limit".to_string(), "100".to_string());

        if let Some(opts) = opts {
            if let Some(s) = opts.search {
                params.insert("q".to_string(), s);
            }
            if let Some(a) = opts.action {
                params.insert("action".to_string(), a);
            }
            if let Some(e) = opts.entity {
                params.insert("entity".to_string(), e);
            }
            if let Some(tags) = opts.tags {
                params.insert("tags".to_string(), tags.join(","));
            }
            if let Some(since) = opts.since {
                params.insert("since".to_string(), since.to_rfc3339());
            }
            if let Some(until) = opts.until {
                params.insert("until".to_string(), until.to_rfc3339());
            }
            if let Some(l) = opts.limit {
                params.insert("limit".to_string(), l.to_string());
            }
            if let Some(o) = opts.offset {
                params.insert("offset".to_string(), o.to_string());
            }
        }

        let response: EntriesResponse = self.client.get("/memory", Some(&params)).await?;
        Ok(response.entries)
    }

    /// Gets a specific memory entry by hash.
    pub async fn get(&self, entry_hash: &str) -> Result<MemoryEntry, Error> {
        self.client
            .get(&format!("/memory/{}", entry_hash), None)
            .await
    }

    /// Gets recent memory entries.
    pub async fn recent(&self, limit: Option<i32>) -> Result<Vec<MemoryEntry>, Error> {
        self.query(Some(MemoryQueryOptions {
            limit: Some(limit.unwrap_or(50)),
            ..Default::default()
        }))
        .await
    }

    /// Gets agent state.
    pub async fn agent_state(
        &self,
        agent_id: &str,
    ) -> Result<HashMap<String, serde_json::Value>, Error> {
        self.client
            .get(&format!("/memory/agents/{}/state", agent_id), None)
            .await
    }

    /// Syncs agent state.
    pub async fn sync_state(
        &self,
        agent_id: &str,
        state: HashMap<String, serde_json::Value>,
    ) -> Result<(), Error> {
        let _: serde_json::Value = self
            .client
            .post(&format!("/memory/agents/{}/state", agent_id), &state)
            .await?;
        Ok(())
    }

    /// Broadcasts a message.
    pub async fn broadcast(&self, msg_type: &str, payload: &str) -> Result<String, Error> {
        let body = serde_json::json!({
            "type": msg_type,
            "payload": payload
        });
        let response: BroadcastResponse = self.client.post("/memory/broadcast", &body).await?;
        Ok(response.broadcast_id)
    }

    /// Creates a TIL (Today I Learned) entry.
    pub async fn til(&self, category: &str, learning: &str) -> Result<MemoryEntry, Error> {
        self.log(LogMemoryOptions {
            action: "til".to_string(),
            entity: category.to_string(),
            details: Some(learning.to_string()),
            tags: Some(vec!["til".to_string(), category.to_string()]),
            metadata: None,
        })
        .await
    }

    /// Gets memory statistics.
    pub async fn stats(&self) -> Result<Stats, Error> {
        self.client.get("/memory/stats", None).await
    }

    /// Verifies hash chain integrity.
    pub async fn verify_chain(&self, start_hash: Option<&str>) -> Result<VerifyChainResult, Error> {
        let params = match start_hash {
            Some(h) => {
                let mut p = HashMap::new();
                p.insert("start".to_string(), h.to_string());
                Some(p)
            }
            None => None,
        };

        self.client.get("/memory/verify", params.as_ref()).await
    }
}
